File: 1944-number-of-visible-people-in-a-queue/1944-number-of-visible-people-in-a-queue.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> ans(n);
        stack<int> st;
        for (int i = n-1; i >= 0; i--) { // 뒤에서
            while (!st.empty() && heights[i] > st.top()) {
                st.pop();
                ans[i]++; // [i+1,] 에서 더 작은 것들
            }
            if (!st.empty())
                ans[i]++; // 스택보다 작음 == 다음 것만 봄
            st.push(heights[i]);
        }
        return ans;
    }
};

File: 1944-number-of-visible-people-in-a-queue/1944-number-of-visible-people-in-a-queue.py
class Solution:
    def canSeePersonsCount(self, heights):
        res = [0] * len(heights)
        stack = []
        for i, v in enumerate(heights):
            while stack and heights[stack[-1]] <= v:
                res[stack.pop()] += 1
            if stack:
                res[stack[-1]] += 1
            stack.append(i)
        return res
            
    # time limit exceeded
    # def canSeePersonsCount(self, heights: List[int]) -> List[int]:
    #     answers = [0] * len(heights)
    #     for i in range(len(heights)):
    #         stack = []
    #         for j in range(i+1, len(heights)):
    #             new = heights[j]
    #             if len(stack) > 0 and stack[-1] >= new:
    #                 continue
    #             stack.append(new)
    #             if new > heights[i]:
    #                 break
    #         answers[i] = len(stack)
    #     return answers
                    